File: app/src/main/java/io/connorwyatt/flashcards/fragments/FlashcardTestFragment.java
package io.connorwyatt.flashcards.fragments;

import android.app.AlertDialog;
import android.app.Fragment;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.android.internal.util.Predicate;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import io.connorwyatt.flashcards.R;
import io.connorwyatt.flashcards.adapters.FlashcardTestPagerAdapter;
import io.connorwyatt.flashcards.data.datasources.FlashcardDataSource;
import io.connorwyatt.flashcards.data.entities.Flashcard;
import io.connorwyatt.flashcards.data.entities.FlashcardTest;
import io.connorwyatt.flashcards.interfaces.IPerformanceBreakdown;
import io.connorwyatt.flashcards.utils.ListUtils;
import io.connorwyatt.flashcards.views.directionalviewpager.DirectionalViewPager;
import io.connorwyatt.flashcards.views.progressbar.ProgressBar;

public class FlashcardTestFragment extends Fragment {
    private int initialCount;
    private IPerformanceBreakdown performanceBreakdown = createPerformanceBreakdown();
    private HashMap<Long, FlashcardTest> flashcardTestMap = new HashMap<>();
    private ArrayList<Long> skippedFlashcards = new ArrayList<>();
    private FlashcardTestPagerAdapter flashcardTestPagerAdapter;
    private ProgressBar progressBar;
    private List<Flashcard> flashcards;
    private ArrayList<IPerformanceBreakdown.OnPerformanceBreakdownChangeListener> changeListeners
            = new ArrayList<>();

    private int getCompletedCardsCount() {
        return performanceBreakdown.getRatedTotal() + performanceBreakdown.getSkipCount();
    }

    public IPerformanceBreakdown getPerformanceBreakdown() {
        return performanceBreakdown;
    }

    private void setUpProgressBar(ViewGroup viewGroup) {
        progressBar = (ProgressBar) viewGroup.findViewById(R.id
                .flashcard_test_progress_bar);
        DirectionalViewPager viewPager = (DirectionalViewPager) viewGroup.findViewById(R.id
                .flashcard_test_view_pager);

        viewPager.addOnPageSkipListener(new DirectionalViewPager.OnPageSkipListener() {
            @Override
            public void onPageSkip(Object skippedItem) {
                updateProgressBar();
            }
        });

        updateProgressBar();
    }

    private void setUpViewPager(ViewGroup viewGroup) {
        DirectionalViewPager viewPager = (DirectionalViewPager) viewGroup.findViewById(R.id
                .flashcard_test_view_pager);
        viewPager.setAdapter(flashcardTestPagerAdapter);

        viewPager.setAllowLeftSwipe(false);
        viewPager.addOnPageSkipListener(new DirectionalViewPager.OnPageSkipListener() {
            @Override
            public void onPageSkip(Object skippedItem) {
                Flashcard flashcard = ((Flashcard) skippedItem);

                if (!flashcardTestMap.containsKey(flashcard.getId())) {
                    skippedFlashcards.add(flashcard.getId());
                    String skipMessage = getString(R.string.flashcard_test_skip_toast, flashcard
                            .getTitle());
                    Toast.makeText(getActivity(), skipMessage, Toast.LENGTH_SHORT)
                            .show();
                    dispatchOnPerformanceBreakdownChangeEvent();
                }
            }
        });
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setRetainInstance(true);

        Intent intent = getActivity().getIntent();

        FlashcardDataSource fds = new FlashcardDataSource(getActivity());
        fds.open();
        if (intent.hasExtra(EXTRA_KEYS.CATEGORY_ID)) {
            flashcards = fds.getByCategory(intent.getLongExtra(EXTRA_KEYS.CATEGORY_ID, -1));
        } else {
            flashcards = fds.getAll();
        }
        fds.close();

        initialCount = flashcards.size();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        ViewGroup viewGroup = (ViewGroup) inflater.inflate(R.layout.fragment_flashcard_test,
                container, false);

        flashcardTestPagerAdapter = new FlashcardTestPagerAdapter(getFragmentManager(), flashcards);

        setUpViewPager(viewGroup);

        setUpProgressBar(viewGroup);

        return viewGroup;
    }

    public void onBackPressed(final Runnable runnable) {
        int totalCompleted = performanceBreakdown.getRatedTotal() + performanceBreakdown
                .getSkipCount();
        boolean isComplete = totalCompleted >= initialCount;

        if (isComplete) {
            runnable.run();
        } else {
            new AlertDialog.Builder(getActivity())
                    .setTitle(R.string.flashcard_test_confirmation_title)
                    .setMessage(R.string.flashcard_test_confirmation_message)
                    .setPositiveButton(R.string.flashcard_test_confirmation_yes, new DialogInterface
                            .OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            runnable.run();
                        }
                    })
                    .setNegativeButton(R.string.flashcard_test_confirmation_no, new DialogInterface
                            .OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                        }
                    })
                    .create()
                    .show();
        }
    }

    public void updateFlashcardTest(FlashcardTest flashcardTest) {
        flashcardTestMap.put(flashcardTest.getFlashcardId(), flashcardTest);
        dispatchOnPerformanceBreakdownChangeEvent();
    }

    private void updateProgressBar() {
        double percent = (double) getCompletedCardsCount() / (double) initialCount;

        progressBar.setProgress(percent, true);
    }

    private void dispatchOnPerformanceBreakdownChangeEvent() {
        for (IPerformanceBreakdown.OnPerformanceBreakdownChangeListener changeListener :
                changeListeners) {
            changeListener.onChange();
        }
    }

    private IPerformanceBreakdown createPerformanceBreakdown() {
        return new IPerformanceBreakdown() {
            @Override
            public int getNegativeCount() {
                return getRatingCount(FlashcardTest.Rating.NEGATIVE);
            }

            @Override
            public double getNegativePercent() {
                return getPercentage(getNegativeCount(), getRatedTotal());
            }

            @Override
            public int getNeutralCount() {
                return getRatingCount(FlashcardTest.Rating.NEUTRAL);
            }

            @Override
            public double getNeutralPercent() {
                return getPercentage(getNeutralCount(), getRatedTotal());
            }

            @Override
            public int getPositiveCount() {
                return getRatingCount(FlashcardTest.Rating.POSITIVE);
            }

            @Override
            public double getPositivePercent() {
                return getPercentage(getPositiveCount(), getRatedTotal());
            }

            @Override
            public int getRatedTotal() {
                return flashcardTestMap.size();
            }

            @Override
            public int getSkipCount() {
                return skippedFlashcards.size();
            }

            @Override
            public double getSkipPercent() {
                return getPercentage(getSkipCount(), getTotal());
            }

            @Override
            public int getTotal() {
                return initialCount;
            }

            @Override
            public void addOnPerformanceBreakdownChangeListener
                    (OnPerformanceBreakdownChangeListener listener) {
                changeListeners.add(listener);
            }

            @Override
            public void removeOnPerformanceBreakdownChangeListener
                    (OnPerformanceBreakdownChangeListener listener) {
                changeListeners.remove(listener);
            }

            @Override
            public void clearOnPerformanceBreakdownChangeListener() {
                changeListeners.clear();
            }

            private int getRatingCount(final FlashcardTest.Rating rating) {
                return ListUtils.filter(new ArrayList<>(flashcardTestMap.values()), new
                        Predicate<FlashcardTest>() {
                            @Override
                            public boolean apply(FlashcardTest flashcardTest) {
                                return flashcardTest.getRating() == rating;
                            }
                        }).size();
            }

            private double getPercentage(int count, int total) {
                double percentage = ((double) count / (double) total);

                if (Double.isNaN(percentage)) percentage = 0;

                return percentage;
            }
        };
    }

    public static class EXTRA_KEYS {
        public static String CATEGORY_ID = "CATEGORY_ID";
    }
}

File: app/src/main/kotlin/io/connorwyatt/flashcards/fragments/FlashcardTestFragment.kt
/*
 * Copyright (c) 2016-2017 Connor Wyatt <connorwyatt1@gmail.com>.
 *
 * This file can not be copied and/or distributed without the express permission of Connor Wyatt.
 */

package io.connorwyatt.flashcards.fragments

import android.app.AlertDialog
import android.app.Fragment
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import io.connorwyatt.flashcards.R
import io.connorwyatt.flashcards.activities.FlashcardTestActivity.Companion.Order
import io.connorwyatt.flashcards.activities.FlashcardTestActivity.Companion.Order.RANDOM
import io.connorwyatt.flashcards.activities.FlashcardTestActivity.Companion.Order.WORST_TO_BEST
import io.connorwyatt.flashcards.adapters.FlashcardTestPagerAdapter
import io.connorwyatt.flashcards.data.entities.Flashcard
import io.connorwyatt.flashcards.data.entities.FlashcardTest
import io.connorwyatt.flashcards.data.services.FlashcardService
import io.connorwyatt.flashcards.data.services.FlashcardTestService
import io.connorwyatt.flashcards.data.viewmodels.FlashcardViewModel
import io.connorwyatt.flashcards.data.viewmodels.PerformanceViewModel
import io.connorwyatt.flashcards.enums.Rating
import io.connorwyatt.flashcards.shuffle
import io.connorwyatt.flashcards.views.directionalviewpager.DirectionalViewPager
import io.connorwyatt.flashcards.views.progressbar.ProgressBar
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject

class FlashcardTestFragment : Fragment() {
  lateinit private var viewGroup: ViewGroup
  lateinit private var progressBar: ProgressBar
  private var flashcardTestPagerAdapter: FlashcardTestPagerAdapter? = null
  private var viewModels: List<FlashcardViewModel>? = null
  private val flashcardTests = mutableMapOf<String, FlashcardTest>()
  private var skippedFlashcards = mutableListOf<String>()
  private val performanceSubject = BehaviorSubject.createDefault(getPerformanceViewModel())

  //region Activity

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    retainInstance = true
  }

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup,
                            savedInstanceState: Bundle?): View {
    super.onCreateView(inflater, container, savedInstanceState)

    viewGroup = inflater.inflate(
      R.layout.fragment_flashcard_test, container, false) as ViewGroup

    val tagId = arguments.getString(ArgumentKeys.TAG_ID)
    val order = Order.values()[arguments.getInt(ArgumentKeys.ORDER)]
    val limit = arguments.getInt(ArgumentKeys.LIMIT, -1).takeUnless { it == -1 }

    initialiseUI(tagId, order, limit)

    updateUI(false)

    return viewGroup
  }

  fun onBackPressed(callback: () -> Unit): Unit {
    val totalCompleted = flashcardTests.size + skippedFlashcards.size
    val isComplete = totalCompleted >= viewModels?.size ?: 0

    if (isComplete) {
      callback()
    } else {
      AlertDialog.Builder(activity)
        .setTitle(R.string.flashcard_test_confirmation_title)
        .setMessage(R.string.flashcard_test_confirmation_message)
        .setPositiveButton(R.string.flashcard_test_confirmation_yes) { _, _ -> callback() }
        .setNegativeButton(R.string.flashcard_test_confirmation_no) { _, _ -> }
        .create()
        .show()
    }
  }

  //endregion

  //region Data

  fun getFlashcardFromAdapter(id: String) = flashcardTestPagerAdapter!!.getFlashcardById(id)

  fun rateFlashcard(viewModel: FlashcardViewModel, rating: Rating): Observable<FlashcardTest> {
    val flashcardTest = FlashcardTest(null)

    val id = viewModel.flashcard.id!!

    flashcardTest.relationships.setRelationships("flashcard", listOf(id))

    flashcardTest.rating = rating

    flashcardTests.put(id, flashcardTest)

    updateUI()

    performanceSubject.onNext(getPerformanceViewModel())

    return saveFlashcardTest(flashcardTest)
  }

  fun getPerformanceObservable(): Observable<PerformanceViewModel> = performanceSubject

  private fun getData(tagId: String?, order: Order, limit: Int?): Observable<List<FlashcardViewModel>> {
    val flashcards = if (tagId !== null)
      FlashcardService.getByTag(tagId)
    else
      FlashcardService.getAll()

    val enhancedFlashcards = flashcards.flatMap {
      val observables = it.map { FlashcardViewModel.getFromFlashcard(it, includeRating = order == WORST_TO_BEST) }

      return@flatMap Observable.combineLatest(observables) {
        val viewModels = it.filterIsInstance(FlashcardViewModel::class.java)

        return@combineLatest viewModels
      }
    }

    val sortedFlashcards = enhancedFlashcards.map {
      when (order) {
        RANDOM -> it.shuffle()
        WORST_TO_BEST -> it.sortedBy { it.averageRating }
        else -> it
      }
    }

    val limitedFlashcards = limit?.let { sortedFlashcards.map { it.subList(0, limit) } } ?: sortedFlashcards

    return limitedFlashcards
  }

  private fun saveFlashcardTest(flashcardTest: FlashcardTest): Observable<FlashcardTest> {
    return FlashcardTestService.save(flashcardTest)
  }

  private fun getPerformanceViewModel(): PerformanceViewModel {
    val ratings = flashcardTests.mapNotNull { it.value.rating }

    return PerformanceViewModel(ratings, viewModels?.size ?: 0)
  }

  //endregion

  //region UI

  private fun initialiseUI(tagId: String?, order: Order, limit: Int?): Unit {
    initialisePager(tagId, order, limit)

    initialiseProgressBar()
  }

  private fun initialisePager(tagId: String?, order: Order, limit: Int?): Unit {
    flashcardTestPagerAdapter = FlashcardTestPagerAdapter(fragmentManager)

    if (viewModels == null) {
      getData(tagId, order, limit).subscribe {
        viewModels = it

        flashcardTestPagerAdapter!!.setData(it)

        updateUI()

        performanceSubject.onNext(getPerformanceViewModel())
      }
    } else {
      flashcardTestPagerAdapter!!.setData(viewModels!!)

      performanceSubject.onNext(getPerformanceViewModel())
    }

    val viewPager =
      viewGroup.findViewById(R.id.flashcard_test_view_pager) as DirectionalViewPager

    viewPager.adapter = flashcardTestPagerAdapter
    viewPager.allowLeftSwipe = false

    viewPager.addOnPageSkipListener {
      val viewModel = it as FlashcardViewModel
      val flashcardId = viewModel.flashcard.id!!

      if (!flashcardTests.containsKey(flashcardId)) {
        skippedFlashcards.add(flashcardId)

        val skipMessage = getString(R.string.flashcard_test_skip_toast, viewModel.flashcard.title)
        Toast.makeText(activity, skipMessage, Toast.LENGTH_SHORT)
          .show()
      }

      updateUI()

      performanceSubject.onNext(getPerformanceViewModel())
    }
  }

  private fun initialiseProgressBar(): Unit {
    progressBar = viewGroup.findViewById(R.id.flashcard_test_progress_bar) as ProgressBar
  }

  private fun updateUI(animate: Boolean = true): Unit {
    updateProgressBar(animate)
  }

  private fun updateProgressBar(animate: Boolean): Unit {
    val ratedCards = flashcardTests.size.toDouble()
    val skippedCards = skippedFlashcards.size.toDouble()
    val totalCards = viewModels?.size?.toDouble()

    totalCards?.let {
      progressBar.setProgress((ratedCards + skippedCards) / totalCards, animate)
    }
  }

  //endregion

  companion object {
    object ArgumentKeys {
      val TAG_ID = "TAG_ID"
      val ORDER = "ORDER"
      val LIMIT = "LIMIT"
    }
  }
}
